main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -442,12 +443,50 @@ object Verifier {
       val t = visitExpr(exp)
       checkEq(tpe, t, loc)
 
-    case Expr.TryWith(_, _, _, _, tpe, _, _) =>
-      // TODO: VERIFIER: Add support for TryWith.
-      tpe
+    case Expr.TryWith(exp, effUse, rules, ct, tpe, purity, loc) =>
+      val exptype = visitExpr(exp) match {
+        case MonoType.Arrow(List(u), t) =>
+          check(expected = MonoType.Unit)(actual = u, loc)
+          t
+        case e => failMismatchedShape(e, "Arrow with one parameter", loc)
+      }
+
+      val effect = root.effects.getOrElse(effUse.sym,
+        throw InternalCompilerException(s"Unknown effect sym: '${effUse.sym}'", loc))
+      val ops = effect.ops.map(op => op.sym -> op).toMap
+
+      for (rule <- rules) {
+        val ruletype = visitExpr(rule.exp)
+        val op = ops.getOrElse(rule.op.sym,
+          throw InternalCompilerException(s"Unknown operation sym: '${rule.op.sym}'", loc))
+
+        val params = op.fparams.map(_.tpe)
+        val resumptionType = MonoType.Arrow(List(op.tpe), exptype)
+        val signature = MonoType.Arrow(params :+ resumptionType, exptype)
+
+        checkEq(ruletype, signature, loc)
+      }
+
+      checkEq(tpe, exptype, loc)
+
+    case Expr.Do(opUse, exps, tpe, purity, loc) =>
+      val ts = exps.map(visitExpr)
+      val eff = root.effects.getOrElse(opUse.sym.eff,
+        throw InternalCompilerException(s"Unknown effect sym: '${opUse.sym.eff}'", loc))
+      val op = eff.ops.find(_.sym == opUse.sym)
+        .getOrElse(throw InternalCompilerException(s"Unknown operation sym: '${opUse.sym}'", loc))
+
+      val oprestype = op.tpe match {
+        case MonoType.Void => tpe // should match any return type
+        case t => t
+      }
+
+      val sig = MonoType.Arrow(ts, tpe)
+      val opsig = MonoType.Arrow(
```

Comment:
Deadcode? How are these locals used?

Reply:
They are only used on line 489. Should they be removed and just inlined into the checkEq call?

Reply:
No, its my bad :)